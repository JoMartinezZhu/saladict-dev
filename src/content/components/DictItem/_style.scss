.panel-DictItem {
  position: relative;

  &:first-child {
    .panel-DictItem_Header {
      border-top: none;
    }
  }
}

.panel-DictItem_Header {
  display: flex;
  align-items: flex-start;
  position: sticky;
  top: 0;
  z-index: 100;
  height: 20px;
  border-top: 1px #ddd solid;
  background: #fff;

  &::after {
    content: '';
    position: absolute;
    top: 20px - 1px;
    left: 0;
    width: 100%;
    height: 10px;
    background: linear-gradient(#fff 5%, rgba(255, 255, 255, 0.3) 60%, transparent 100%);
  }
}

.panel-DictItem_Logo {
  width: 20px;
  height: 20px;
  margin-top: -1px;
  user-select: none;
}

.panel-DictItem_Title {
  margin: 0;
  padding: 0 3px;
  line-height: 20px;
  font-size: 12px;
  font-weight: normal;

  a {
    color: #444;
    text-decoration: none;
  }
}

.panel-DictItem_Body {
  overflow: hidden;
  position: relative;
  padding-left: 10px;
  padding-right: 5px;
  font-size: 12px;
  line-height: 1.6;
  color: #333;
}

.panel-DictItem_BodyMesure {
  // clear margin collapsing which seems to causing flickering
  // when changing the parent height
  overflow: hidden;

  > *:first-child {
    margin-top: 10px !important;
  }

  > *:last-child {
    margin-bottom: 10px !important;
  }
}

.panel-DictItem_Loader {
  align-self: center;
  width: 120px;
  height: 10px;
  user-select: none;
}

.panel-DictItem_Loader_Ball {
  width: 10px;
  height: 10px;
  fill: orange;

  &:nth-child(2) {
    transform: translateX(15px);
  }
  &:nth-child(3) {
    transform: translateX(30px);
  }
  &:nth-child(4) {
    transform: translateX(45px);
  }
  &:nth-child(5) {
    transform: translateX(60px);
  }
}

.panel-DictItem_FoldArrowBtn {
  width: 20px - 1px;
  height: 20px - 1px;
  overflow: hidden;
  background: none;
  border: none;
  margin: 0 0 0 auto;
  padding: 0;
  cursor: pointer;
  user-select: none;

  &:hover {
    outline: none;
  }
}

.panel-DictItem_FoldArrow {
  fill: #000;
  width: 18px;
  height: 18px;
  padding: 3px;
}

// transform svg path instead of svg itself to
// prevent iframe flickering bug in Chrome
.panel-DictItem_FoldArrowPath {
  transform-origin: center;
  &.isActive {
    transform: rotate(-90deg);
  }
}

.panel-DictItem_FoldMask {
  position: absolute;
  left: 0;
  bottom: 0;
  width: 100%;
  height: 50px;
  padding: 0;
  border: none;
  background: linear-gradient(transparent 40%, rgba(255, 255, 255, 0.3) 60%, rgba(255, 255, 255, 0.9) 100%);
  cursor: pointer;
  user-select: none;

  &::after {
    content: '';
    position: absolute;
    top: 100%;
    left: 0;
    width: 100%;
    height: 100%;
    background: linear-gradient(transparent 40%, rgba(225, 225, 225, .2) 60%, rgba(225, 225, 225, .7) 100%);
  }

  &:hover {
    outline: none;

    &::after {
      /**
       * use top instead of opacity to avoid GUP acceleration
       * which might casue iframe flickering in Chrome 66
       * @todo add back after Chrome 68
       */
      top: 0;
    }
  }
}

.panel-DictItem_FoldMaskArrow {
  position: absolute;
  z-index: 10;
  bottom: 0;
  left: 0;
  right: 0;
  margin: 0 auto;
  width: 15px;
  height: 15px;
  fill: #000;
}

/*-----------------------------------------------*\
    States
\*-----------------------------------------------*/

.isAnimate {
  .panel-DictItem_Loader_Ball {
    animation: panel-DictItem_Loader-shift 2s linear infinite;

    &:nth-child(2) {
      animation-delay: -0.4s;
    }
    &:nth-child(3) {
      animation-delay: -0.8s;
    }
    &:nth-child(4) {
      animation-delay: -1.2s;
    }
    &:nth-child(5) {
      animation-delay: -1.6s;
    }
  }

  .panel-DictItem_FoldArrowPath {
    transition: transform 400ms;
  }

  .panel-DictItem_FoldMask::after {
    transition: top 400ms;
  }

  .panel-DictItem_Body {
    transition: height 0.4s, opacity 0.4s;
  }
}

@keyframes panel-DictItem_Loader-shift {
    0% { transform: translateX(  0  ); opacity: 0; }
   10% { transform: translateX( 30px); opacity: 1; }
   90% { transform: translateX( 80px); opacity: 1; }
  100% { transform: translateX(110px); opacity: 0; }
}
