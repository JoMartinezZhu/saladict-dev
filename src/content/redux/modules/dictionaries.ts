import { message } from '@/_helpers/browser-api'
import { DictID, PreloadSource } from '@/app-config'
import isEqual from 'lodash/isEqual'
import { saveWord } from '@/_helpers/record-manager'
import { getDefaultSelectionInfo, SelectionInfo, isSameSelection } from '@/_helpers/selection'
import { MsgType, MsgFetchDictResult, MsgQSPanelSearchText, MsgFetchDictResultResponse, MsgAudioPlay } from '@/typings/message'
import getDefaultProfile from '@/app-config/profiles'
import { DeepReadonly } from '@/typings/helpers'
import { StoreState, DispatcherThunk } from './index'
import { isInNotebook, searchBoxUpdate } from './widget'
import {
  countWords,
  checkSupportedLangs,
} from '@/_helpers/lang-check'
import { MachineTranslateResult } from '@/components/dictionaries/helpers'

const isSaladictOptionsPage = !!window.__SALADICT_OPTIONS_PAGE__
const isSaladictInternalPage = !!window.__SALADICT_INTERNAL_PAGE__
const isSaladictPopupPage = !!window.__SALADICT_POPUP_PAGE__
const isSaladictQuickSearchPage = !!window.__SALADICT_QUICK_SEARCH_PAGE__
const isSaladictPDFPage = !!window.__SALADICT_PDF_PAGE__

const isStandalonePage = isSaladictPopupPage || isSaladictQuickSearchPage
const isNoSearchHistoryPage = isSaladictInternalPage && !isStandalonePage

/*-----------------------------------------------*\
    Action Type
\*-----------------------------------------------*/

export const enum ActionType {
  NEW_CONFIG = 'dicts/NEW_CONFIG',
  SEARCH_START = 'dicts/SEARCH_START',
  SEARCH_END = 'dicts/SEARCH_END',
  RESTORE = 'dicts/RESTORE',
  ADD_HISTORY = 'dicts/ADD_HISTORY',
}

/*-----------------------------------------------*\
    Payload
\*-----------------------------------------------*/

interface DictionariesPayload {
  [ActionType.NEW_CONFIG]: undefined
  [ActionType.RESTORE]: undefined
  [ActionType.ADD_HISTORY]: SelectionInfo
  [ActionType.SEARCH_START]: {
    toOnhold: DictID[]
    toStart: DictID[]
    toActive?: DictID[]
    info: SelectionInfo
  }
  [ActionType.SEARCH_END]: {
    id: DictID
    info: SelectionInfo
    result: any
  }
}

/*-----------------------------------------------*\
    State
\*-----------------------------------------------*/

export const enum SearchStatus {
  OnHold,
  Searching,
  Finished,
}

type DictState = {
  readonly searchStatus: SearchStatus
  readonly searchResult: any
}

export type DictionariesState = {
  readonly dictionaries: {
    readonly selected: DeepReadonly<DictID[]>
    readonly active: DictID[]
    readonly dicts: {
      readonly [k in DictID]?: DictState
    }
    // 0 is the latest
    readonly searchHistory: SelectionInfo[]
  }
}

const defaultProfile = getDefaultProfile()

export const initState: DictionariesState = {
  dictionaries: {
    selected: defaultProfile.dicts.selected,
    active: [],
    dicts: defaultProfile.dicts.selected
      .reduce((state, id) => {
        state[id] = {
          searchStatus: SearchStatus.OnHold,
          searchResult: null,
        }
        return state
      }, {}),
    searchHistory: [],
  }
}

/*-----------------------------------------------*\
    Reducer Object
\*-----------------------------------------------*/

type DictsReducer = {
  [k in ActionType]: (state: StoreState, payload: DictionariesPayload[k]) => StoreState
}

export const reducer: DictsReducer = {
  [ActionType.NEW_CONFIG] (state) {
    const { dictionaries } = state
    const { selected } = state.config.dicts

    if (isEqual(selected, dictionaries.selected)) {
      return state
    }

    return {
      ...state,
      dictionaries: {
        ...dictionaries,
        selected: selected.slice(),
        active: dictionaries.active.filter(id => selected.indexOf(id) !== -1),
        dicts: selected.reduce((newState, id) => {
          newState[id] = dictionaries.dicts[id] || {
            searchStatus: SearchStatus.OnHold,
            searchResult: null,
          }
          return newState
        }, {}),
      }
    }
  },
  [ActionType.RESTORE] (state) {
    const { dictionaries } = state
    return {
      ...state,
      dictionaries: {
        ...dictionaries,
        active: [],
        dicts: Object.keys(dictionaries.dicts).reduce((newDicts, id) => {
          newDicts[id] =
            dictionaries.dicts[id].searchStatus === SearchStatus.OnHold
              ? dictionaries.dicts[id]
              : {
                searchStatus: SearchStatus.OnHold,
                searchResult: null,
              }
          return newDicts
        }, {})
      }
    }
  },
  [ActionType.SEARCH_START] (state, { toStart, toOnhold, toActive, info }) {
    const { dictionaries, widget } = state
    const searchBoxIndex = widget.searchBox.index || 0
    const dicts = { ...dictionaries.dicts }
    toOnhold.forEach(id => {
      if (dicts[id]) {
        dicts[id] = {
          ...dicts[id],
          searchStatus: SearchStatus.OnHold,
          searchResult: null,
        }
      }
    })
    toStart.forEach(id => {
      if (dicts[id]) {
        dicts[id] = {
          ...dicts[id],
          searchStatus: SearchStatus.Searching,
          searchResult: null,
        }
      }
    })

    return {
      ...state,
      dictionaries: {
        ...dictionaries,
        active: toActive || dictionaries.active,
        searchHistory: info === dictionaries.searchHistory[searchBoxIndex]
          ? dictionaries.searchHistory
          // don't create history for same info
          : isSameSelection(info, dictionaries.searchHistory[0])
            ? [info, ...dictionaries.searchHistory.slice(1)]
            : [info, ...dictionaries.searchHistory],
        dicts,
      }
    }
  },
  [ActionType.SEARCH_END] (state, { id, info, result }) {
    const { dictionaries, widget } = state

    if (!isSameSelection(info, dictionaries.searchHistory[widget.searchBox.index || 0])) {
      // ignore the outdated selection
      return state
    }

    return {
      ...state,
      dictionaries: {
        ...dictionaries,
        dicts: {
          ...dictionaries.dicts,
          [id]: { ...dictionaries[id], searchStatus: SearchStatus.Finished, searchResult: result }
        }
      }
    }
  },
  [ActionType.ADD_HISTORY] (state, info) {
    const history = state.dictionaries.searchHistory
    return {
      ...state,
      dictionaries: {
        ...state.dictionaries,
        searchHistory: [info, ...history]
      }
    }
  }
}

/*-----------------------------------------------*\
    Action Creators
\*-----------------------------------------------*/

interface Action<T extends ActionType> {
  type: T,
  payload?: DictionariesPayload[T]
}

export function newConfig (): Action<ActionType.NEW_CONFIG> {
  return ({ type: ActionType.NEW_CONFIG })
}

export function restoreDicts (): Action<ActionType.RESTORE> {
  return ({ type: ActionType.RESTORE })
}

/** Search all selected dicts if id is not provided */
export function searchStart (
  payload: DictionariesPayload[ActionType.SEARCH_START]
): Action<ActionType.SEARCH_START> {
  return ({ type: ActionType.SEARCH_START, payload })
}

export function searchEnd (
  payload: DictionariesPayload[ActionType.SEARCH_END]
): Action<ActionType.SEARCH_END> {
  return ({ type: ActionType.SEARCH_END, payload })
}

export function addSearchHistory (
  payload: DictionariesPayload[ActionType.ADD_HISTORY]
): Action<ActionType.ADD_HISTORY> {
  return ({ type: ActionType.ADD_HISTORY, payload })
}

/*-----------------------------------------------*\
    Side Effects
\*-----------------------------------------------*/

export function startUpAction (): DispatcherThunk {
  return (dispatch, getState) => {
    if (isSaladictPopupPage) {
      const { baPreload, baAuto } = getState().config
      dispatch(summonedPanelInit(baPreload, baAuto, 'popup'))
    } else if (isSaladictQuickSearchPage) {
      /** From other tabs */
      message.addListener<MsgQSPanelSearchText>(MsgType.QSPanelSearchText, ({ info }) => {
        dispatch(searchText({ info }))
        // focus standalone panel
        message.send({ type: MsgType.OpenQSPanel })
      })
    }
  }
}

/**
 * Search all selected dicts if id is not provided.
 * Use last selection if info is not provided.
 */
export function searchText (
  arg?: { id?: DictID, info?: SelectionInfo, payload?: { [index: string]: any } }
): DispatcherThunk {
  return (dispatch, getState) => {
    const state = getState()
    const searchBoxIndex = state.widget.searchBox.index || 0
    const info = arg && arg.info
      ? arg.info
      : state.dictionaries.searchHistory[searchBoxIndex]

    // try to unfold a dict when the panel first popup
    if (!info || !info.text) { return }

    if (isSaladictOptionsPage) {
      window.__SALADICT_LAST_SEARCH__ = info.text
    }

    dispatch(isInNotebook(info))

    const requestID = arg && arg.id

    // search one dict, when user clicks the unfold arrow
    if (requestID) {
      dispatch(searchStart({ toStart: [requestID], toOnhold: [], info }))
      doSearch(requestID)
      return
    }

    const { selected: selectedDicts, all: allDicts } = state.config.dicts
    /** should display */
    const toActive: DictID[] = []
    /** should start searching */
    const toStart: DictID[] = []
    /** reset to folded state */
    const toOnhold: DictID[] = []

    selectedDicts.forEach(id => {
      const dict = allDicts[id]
      let isValidSelection = checkSupportedLangs(dict.selectionLang, info.text)

      if (isValidSelection) {
        const wordCount = countWords(info.text)
        const { min, max } = dict.selectionWC
        isValidSelection = wordCount >= min && wordCount <= max
      }

      if (isValidSelection) {
        toActive.push(id)
      }

      if (isValidSelection && checkSupportedLangs(dict.defaultUnfold, info.text)) {
        toStart.push(id)
      } else {
        toOnhold.push(id)
      }
    })

    if (!isNoSearchHistoryPage &&
        state.config.searhHistory &&
        (!browser.extension.inIncognitoContext || state.config.searhHistoryInco) &&
        !isSameSelection(state.dictionaries.searchHistory[0], info)
    ) {
      saveWord('history', info)
    }

    dispatch(searchStart({ toStart, toOnhold, toActive, info }))
    if (arg && arg.info) {
      // Reset index after search start. Index is useful.
      dispatch(searchBoxUpdate({ text: info.text, index: 0 }))
    }

    const pSearchResponses = toStart.map(doSearch)

    const { cn, en, machine } = state.config.autopron

    // dict with auto pronunciation but not searching
    if (cn.dict && !toStart.includes(cn.dict)) {
      pSearchResponses.push(requestDictResult(cn.dict))
    }
    if (en.dict && !toStart.includes(en.dict)) {
      pSearchResponses.push(requestDictResult(en.dict))
    }
    if (machine.dict && !toStart.includes(machine.dict)) {
      pSearchResponses.push(requestDictResult(machine.dict))
    }

    // handle auto pronunciation
    let hasPlayed = false
    for (const pSearchResponse of pSearchResponses) {
      pSearchResponse.then(({ id, result, audio }) => {
        if (hasPlayed) { return }

        if (audio) {
          if (id === cn.dict && audio.py) {
            message.send<MsgAudioPlay>({ type: MsgType.PlayAudio, src: audio.py })
            hasPlayed = true
            return
          }

          if (id === en.dict) {
            const src = en.accent === 'us'
              ? audio!.us || audio!.uk
              : audio!.uk || audio!.us
            if (src) {
              message.send<MsgAudioPlay>({ type: MsgType.PlayAudio, src })
              hasPlayed = true
              return
            }
          }
        }

        if (id === machine.dict) {
          const src = (result as MachineTranslateResult<DictID>)[machine.src].audio
          if (src) {
            message.send<MsgAudioPlay>({ type: MsgType.PlayAudio, src })
            hasPlayed = true
            return
          }
        }
      })
    }

    function requestDictResult (id: DictID): Promise<MsgFetchDictResultResponse<any>> {
      return message.send<MsgFetchDictResult>({
        type: MsgType.FetchDictResult,
        id,
        text: info.text,
        payload: arg && arg.payload
          ? { isPDF: isSaladictPDFPage, ...arg.payload }
          : { isPDF: isSaladictPDFPage },
      })
    }

    function doSearch (id: DictID): Promise<MsgFetchDictResultResponse<any>> {
      return requestDictResult(id)
        .then(response => {
          dispatch(searchEnd({ id, info, result: response.result }))
          return response
        })
        .catch(() => {
          dispatch(searchEnd({ id, info , result: null }))
          return { id, result: null }
        })
    }
  }
}

export function summonedPanelInit (
  preload: PreloadSource,
  autoSearch: boolean,
  // quick-search could be turned off so this argument is needed
  standalone: '' | 'popup' | 'quick-search',
): DispatcherThunk {
  return async (dispatch, getState) => {
    if (!preload) { return }

    const state = getState()

    let info: SelectionInfo | null = null

    try {
      if (preload === 'selection') {
        if (standalone === 'popup') {
          const tab = (await browser.tabs.query({ active: true, currentWindow: true }))[0]
          if (tab && tab.id != null) {
            info = await message.send(tab.id, { type: MsgType.PreloadSelection })
          }
        } else if (standalone === 'quick-search') {
          const infoText = new URL(document.URL).searchParams.get('info')
          if (infoText) {
            try {
              info = JSON.parse(decodeURIComponent(infoText))
            } catch (err) {
              info = null
            }
          }
        } else {
          info = { ...state.selection.selectionInfo }
        }
      } else /* preload === clipboard */ {
        const text = await message.send({ type: MsgType.GetClipboard })
        info = getDefaultSelectionInfo({ text, title: 'From Clipboard' })
      }
    } catch (e) {
      if (process.env.DEV_BUILD) {
        console.warn(e)
      }
    }

    if (info) {
      if (autoSearch && info.text) {
        dispatch(searchText({ info }))
      } else {
        dispatch(restoreDicts())
        dispatch(searchBoxUpdate({ text: info.text, index: 0 }))
      }
    }
  }
}
