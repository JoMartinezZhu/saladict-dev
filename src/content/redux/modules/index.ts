import mergeUniqueObjects from '../utils/merge-unique-objects'

import {
  reducer as configReducer,
  initState as configInitState,
  ActionType as ConfigActionType,
  ConfigState,
} from './config'

import {
  reducer as widgetReducer,
  initState as widgetInitState,
  ActionType as widgetActionType,
  WidgetState,
} from './widget'

import {
  reducer as dictionariesReducer,
  initState as dictionariesInitState,
  ActionType as dictionariesActionType,
  DictionariesState,
} from './dictionaries'

import {
  reducer as selectionReducer,
  initState as selectionInitState,
  ActionType as selectionActionType,
  SelectionState,
} from './selection'

export type StoreState =
  ConfigState &
  WidgetState &
  DictionariesState &
  SelectionState

export type ActionType =
  ConfigActionType |
  widgetActionType |
  dictionariesActionType |
  selectionActionType

/** Thunk dispatcher */
export type Dispatcher = (action: { type: ActionType, payload?: any } | DispatcherThunk) => any

export type DispatcherThunk = (
  dispatch: Dispatcher,
  getState: () => StoreState
) => any

const storeInitState: StoreState = mergeUniqueObjects(
  configInitState,
  widgetInitState,
  dictionariesInitState,
  selectionInitState,
)

const storeReducer = mergeUniqueObjects(
  configReducer,
  widgetReducer,
  dictionariesReducer,
  selectionReducer,
)

export default function reducer (
  state = storeInitState,
  action: { type: ActionType, payload: any },
): StoreState {
  if (typeof storeReducer[action.type] === 'function') {
    // @ts-ignore
    return storeReducer[action.type](state, action.payload)
  }

  if (process.env.NODE_ENV !== 'production') {
    if (!action.type.startsWith('@@')) {
      console.error(`Action ${action.type} doesn't have a handler`)
    }
  }

  return state
}
