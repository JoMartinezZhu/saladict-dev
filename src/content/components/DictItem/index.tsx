import React from 'react'
import { DictID } from '@/app-config'
import { TranslationFunction } from 'i18next'
import { message } from '@/_helpers/browser-api'
import { MsgType, MsgOpenUrl, MsgOpenSrcPage } from '@/typings/message'

import { SearchStatus } from '@/content/redux/modules/dictionaries'
import { SelectionInfo, getDefaultSelectionInfo } from '@/_helpers/selection'
import { Mutable } from '@/typings/helpers'
import { ViewPorps } from '@/components/dictionaries/helpers'

export interface DictItemDispatchers {
  readonly searchText: (arg?: { id?: DictID, info?: SelectionInfo, payload?: { [index: string]: any } }) => any
  readonly updateItemHeight: (id: DictID, height: number) => any
}

export interface DictItemProps extends DictItemDispatchers {
  readonly t: TranslationFunction
  readonly id: DictID
  readonly text: string
  readonly preferredHeight: number
  readonly searchStatus: SearchStatus
  readonly searchResult: any

  readonly fontSize: number
  readonly panelWidth: number
}

export type DictItemState = {
  readonly propsSearchStatus: SearchStatus | null
  readonly propsSearchResult: any
  readonly offsetHeight: number
  readonly visibleHeight: number
  readonly isUnfold: boolean
  readonly hasError: boolean
}

export default class DictItem extends React.PureComponent<DictItemProps, DictItemState> {
  bodyRef = React.createRef<HTMLDivElement>()
  prevItemHeight = 30
  initStyle = { height: 30, opacity: 0 }

  state = {
    /** same as pros */
    propsSearchStatus: null,
    propsSearchResult: null,
    /** real height of the body */
    offsetHeight: 10,
    /** final height, takes unfold and mask into account */
    visibleHeight: 10,
    isUnfold: false,
    hasError: false,
  }

  static getDerivedStateFromProps (
    nextProps: DictItemProps,
    prevState: DictItemState,
  ): Partial<DictItemState> | null {
    const {
      propsSearchResult: lastResult,
      propsSearchStatus: lastStatus,
    } = prevState
    const {
      searchResult: nextResult,
      searchStatus: nextStatus,
    } = nextProps

    if (lastResult === nextResult && lastStatus === nextStatus) {
      return null
    }

    const result: Mutable<Partial<DictItemState>> = {
      propsSearchStatus: nextStatus,
      propsSearchResult: nextResult,
      hasError: false,
    }

    switch (nextStatus) {
      case SearchStatus.OnHold:
      case SearchStatus.Searching:
        result.isUnfold = false
        result.offsetHeight = 0
        result.visibleHeight = 10
        break
      case SearchStatus.Finished:
        result.isUnfold = true
        break
    }

    return result
  }

  blurAfterClick (e) {
    e.currentTarget.blur()
  }

  calcBodyHeight (force?: boolean): Mutable<Partial<DictItemState>> | null {
    if (this.bodyRef.current) {
      const offsetHeight = Math.max(this.bodyRef.current.offsetHeight, 10) || 10
      if (force || this.state.offsetHeight !== offsetHeight) {
        return { offsetHeight, visibleHeight: Math.min(offsetHeight, this.props.preferredHeight) }
      }
    }
    return null
  }

  toggleFolding = () => {
    if (this.props.searchStatus === SearchStatus.Searching) {
      return
    }

    if (this.state.isUnfold) {
      this.setState({ isUnfold: false, visibleHeight: 10 })
    } else {
      if (this.props.searchResult) {
        const update = this.calcBodyHeight(true)
        if (update) {
          update.isUnfold = true
          update.hasError = false
          this.setState(update as DictItemState)
        } else {
          this.setState({ isUnfold: true, hasError: false })
        }
      } else {
        this.props.searchText({ id: this.props.id })
      }
    }
  }

  showFull = e => {
    // always recalculate, in case of img load delay
    const update = this.calcBodyHeight(true)
    if (update) {
      update.visibleHeight = update.offsetHeight
      this.setState(update as DictItemState)
    } else {
      this.setState(state => ({ visibleHeight: state.offsetHeight }))
    }
    e.currentTarget.blur()
  }

  handleDictItemClick = (e: React.MouseEvent<HTMLElement>) => {
    if (e.ctrlKey || e.metaKey || e.altKey) { return }

    // use background script to open new page
    const $target = e.target as HTMLElement
    if (!$target.tagName || !$target.parentElement) { return }
    const $dictItemBody = this.bodyRef.current || null

    for (let el: HTMLElement | null = $target; el && el !== $dictItemBody; el = el.parentElement) {
      if (el.tagName === 'A' || el.getAttribute('role') === 'link') {
        e.preventDefault()
        e.stopPropagation()

        const $a = el as HTMLAnchorElement
        if (/nofollow|noopener|noreferrer/.test($a.rel)) {
          message.send<MsgOpenUrl>({
            type: MsgType.OpenURL,
            url: $a.href,
          })
        } else {
          this.props.searchText({
            info: getDefaultSelectionInfo({
              text: $a.textContent || '',
              title: this.props.t('fromSaladict'),
              favicon: 'https://raw.githubusercontent.com/crimx/ext-saladict/dev/public/static/icon-16.png'
            }),
          })
        }
        return
      }
    }
  }

  handleDictURLClick = (e: React.MouseEvent<HTMLElement>) => {
    e.stopPropagation()
    e.preventDefault()
    message.send<MsgOpenSrcPage>({
      type: MsgType.OpenSrcPage,
      text: this.props.text,
      id: this.props.id,
    }).catch(e => {/* */})
  }

  handleRecalcBodyHeight = () => {
    setTimeout(() => {
      if (this.bodyRef.current) {
        const offsetHeight = Math.max(this.bodyRef.current.offsetHeight, 10) || 10
        this.setState({
          offsetHeight,
          visibleHeight: offsetHeight,
          isUnfold: true,
          hasError: false,
        })
      }
    }, 0)
  }

  componentDidUpdate (prevProps: DictItemProps) {
    if (this.props.searchStatus === SearchStatus.Finished &&
        this.props.searchStatus !== prevProps.searchStatus) {
      const update = this.calcBodyHeight()
      if (update) { this.setState(update as DictItemState) }
    }

    this.props.updateItemHeight(
      this.props.id,
      this.state.visibleHeight + 20,
    )
  }

  componentDidMount () {
    this.props.updateItemHeight(
      this.props.id,
      this.state.visibleHeight + 20,
    )
  }

  componentDidCatch (error, info) {
    if (process.env.NODE_ENV !== 'production') {
      console.warn(`Dict Component ${this.props.id} has error: `, error, info)
    }
    this.setState({
      isUnfold: false,
      visibleHeight: 10,
      hasError: true,
    })
  }

  render () {
    const {
      t,
      id,
      fontSize,
      searchStatus,
      searchResult,
      searchText,
    } = this.props

    const {
      offsetHeight,
      visibleHeight,
      isUnfold,
      hasError,
    } = this.state

    return (
      <section className='panel-DictItem'
        onClick={this.handleDictItemClick}
      >
        <header className='panel-DictItem_Header' onClick={this.toggleFolding}>
          <img className='panel-DictItem_Logo' src={require('@/components/dictionaries/' + id + '/favicon.png')} alt='dict logo' />
          <h1 className='panel-DictItem_Title'>
            <a href='#' onClick={this.handleDictURLClick}>{t(`dict:${id}`)}</a>
          </h1>
          {searchStatus === SearchStatus.Searching && !hasError &&
            <svg className='panel-DictItem_Loader' width='120' height='10' viewBox='0 0 120 10' xmlns='http://www.w3.org/2000/svg'>
              <circle className='panel-DictItem_Loader_Ball' cx='5' cy='5' r='5' />
              <circle className='panel-DictItem_Loader_Ball' cx='5' cy='5' r='5' />
              <circle className='panel-DictItem_Loader_Ball' cx='5' cy='5' r='5' />
              <circle className='panel-DictItem_Loader_Ball' cx='5' cy='5' r='5' />
              <circle className='panel-DictItem_Loader_Ball' cx='5' cy='5' r='5' />
            </svg>
          }
          <button className={'panel-DictItem_FoldArrowBtn'} onClick={this.blurAfterClick}>
            <svg className='panel-DictItem_FoldArrow' width='18' height='18' viewBox='0 0 59.414 59.414' xmlns='http://www.w3.org/2000/svg'>
              <path className={`panel-DictItem_FoldArrowPath ${isUnfold ? 'isActive' : ''}`} d='M43.854 59.414L14.146 29.707 43.854 0l1.414 1.414-28.293 28.293L45.268 58' />
            </svg>
          </button>
        </header>
        <div className='panel-DictItem_Body'
          key={id}
          style={{ fontSize, height: visibleHeight }}
        >
          <article ref={this.bodyRef} className='panel-DictItem_BodyMesure' style={{ opacity: isUnfold ? 1 : 0 }}>
            {searchResult && !hasError &&
              React.createElement<ViewPorps<any>>(
                require('@/components/dictionaries/' + id + '/View.tsx').default,
                {
                  t,
                  result: searchResult,
                  searchText,
                  recalcBodyHeight: this.handleRecalcBodyHeight,
                }
              )
            }
          </article>
          {isUnfold && searchResult && visibleHeight < offsetHeight &&
            <button
              className={'panel-DictItem_FoldMask'}
              onClick={this.showFull}
            >
              <svg className='panel-DictItem_FoldMaskArrow' width='15' height='15' viewBox='0 0 59.414 59.414' xmlns='http://www.w3.org/2000/svg'>
                <path d='M58 14.146L29.707 42.44 1.414 14.145 0 15.56 29.707 45.27 59.414 15.56' />
              </svg>
            </button>
          }
        </div>
      </section>
    )
  }
}
