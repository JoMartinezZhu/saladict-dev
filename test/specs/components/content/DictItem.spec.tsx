import React from 'react'
import { mount, shallow } from 'enzyme'
import { SearchStatus } from '@/content/redux/modules/dictionaries'
import DictItem, { DictItemProps } from '@/content/components/DictItem'
import { TranslationFunction } from 'i18next'
import _ from 'lodash'

const mockDict = jest.fn(({ result }) => result ? 'rendered' : null)

jest.mock('@/components/dictionaries/bing/View.tsx', () => {
  return {
    default: mockDict
  }
})

type TransProps = DictItemProps & { t: TranslationFunction }

describe('Component/content/DictItem', () => {
  beforeEach(() => {
    mockDict.mockClear()
  })
  it('should render pending correctly', () => {
    const props: TransProps = {
      t: _.identity,
      id: 'bing',
      text: 'string',
      dictURL: 'https://google.com',
      preferredHeight: 100,
      searchStatus: SearchStatus.OnHold,
      searchResult: null,

      fontSize: 14,
      panelWidth: 400,
      updateItemHeight: _.noop,
      searchText: _.noop

    }
    expect(shallow(<DictItem {...props} />)).toMatchSnapshot()
  })

  it('should render result correctly', () => {
    const props: TransProps = {
      t: _.identity,
      id: 'bing',
      text: 'string',
      dictURL: 'https://google.com',
      fontSize: 14,
      searchResult: null,
      preferredHeight: 100,
      panelWidth: 400,
      searchStatus: SearchStatus.Searching,
      updateItemHeight: _.noop,
      searchText: _.noop
    }
    const mounted = mount(<DictItem {...props} />)
    expect(mounted.state('isUnfold')).toBe(false)

    mounted.setProps({ ...props, searchResult: 'result1', searchStatus: SearchStatus.Finished })
    expect(mounted.state('isUnfold')).toBe(true)
    setTimeout(() => {
      expect(mockDict).toHaveBeenCalledWith(expect.objectContaining({ result: 'result1' }))
    }, 0)
  })

  it('toggling unfold', () => {
    const props: TransProps = {
      t: _.identity,
      id: 'bing',
      text: 'string',
      dictURL: 'https://google.com',
      fontSize: 14,
      searchResult: null,
      preferredHeight: 100,
      panelWidth: 400,
      searchStatus: SearchStatus.Searching,
      updateItemHeight: _.noop,
      searchText: jest.fn()
    }
    const shallowed = shallow(<DictItem {...props} />)
    expect(shallowed.state('isUnfold')).toBe(false)

    shallowed.setProps({ ...props, searchStatus: SearchStatus.Finished })
    expect(shallowed.state('isUnfold')).toBe(true)

    shallowed.find('header').simulate('click')
    expect(shallowed.state('isUnfold')).toBe(false)

    // request searching
    expect(props.searchText).toHaveBeenCalledTimes(0)
    shallowed.find('header').simulate('click')
    expect(shallowed.state('isUnfold')).toBe(false)
    expect(props.searchText).toHaveBeenCalledTimes(1)
  })
})
