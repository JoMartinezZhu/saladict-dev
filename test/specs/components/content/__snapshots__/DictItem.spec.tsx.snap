// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component/content/DictItem should render pending correctly 1`] = `
<section
  className="panel-DictItem"
  onClick={[Function]}
>
  <header
    className="panel-DictItem_Header"
    onClick={[Function]}
  >
    <img
      alt="dict logo"
      className="panel-DictItem_Logo"
      src="favicon.png"
    />
    <h1
      className="panel-DictItem_Title"
    >
      <a
        href="#"
        onClick={[Function]}
      >
        dict:bing
      </a>
    </h1>
    <button
      className="panel-DictItem_FoldArrowBtn"
      onClick={[Function]}
    >
      <svg
        className="panel-DictItem_FoldArrow"
        height="18"
        viewBox="0 0 59.414 59.414"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          className="panel-DictItem_FoldArrowPath "
          d="M43.854 59.414L14.146 29.707 43.854 0l1.414 1.414-28.293 28.293L45.268 58"
        />
      </svg>
    </button>
  </header>
  <div
    className="panel-DictItem_Body"
    key="bing"
    style={
      Object {
        "fontSize": 14,
        "height": 10,
      }
    }
  >
    <article
      className="panel-DictItem_BodyMesure"
      style={
        Object {
          "opacity": 0,
        }
      }
    />
  </div>
</section>
`;
